### ISSUE TITLE
mkvmerge GUI Header Editor Adds Fields Not Previously Present

### ISSUE THREAD

**Issue by enonu** [mbunkus/mkvtoolnix#1181]
I tried the mkvmerge GUI editor for the first time last night, and using mkvinfo to double check the results, I noticed several extra fields added per-track. Some are probably pretty innocuous, but others like "Interlaced" seem questionable (even though if probably correct). My expected behavior is that if I change one field in the UI, precisely one field is changed.

Below is a diff from mkvinfo on the original file and on edited file. The MKV was generated by the makemkv program reading a BluRay called OLDBOY. Note that I did a simple edit to fix the name of an audio track (Oldboy -> Stereo), but all other changes aren't mine:

```
$ diff OLDBOY-orig.mkv.info OLDBOY-edited.mkv.info 
11c11
< |+ EbmlVoid (size: 1021)
---
> |+ EbmlVoid (size: 13)
18a19
> | + Title: OLDBOY
34a36,40
> |   + Interlaced: 0
> |  + Enabled: 1
> |  + Forced flag: 0
> |  + Max BlockAddition ID: 0
> |  + Codec decode all: 1
47a54,58
> |  + Enabled: 1
> |  + Forced flag: 0
> |  + MinCache: 0
> |  + Max BlockAddition ID: 0
> |  + Codec decode all: 1
60a72,76
> |  + Enabled: 1
> |  + Forced flag: 0
> |  + MinCache: 0
> |  + Max BlockAddition ID: 0
> |  + Codec decode all: 1
74a91,95
> |  + Enabled: 1
> |  + Forced flag: 0
> |  + MinCache: 0
> |  + Max BlockAddition ID: 0
> |  + Codec decode all: 1
87a109,113
> |  + Enabled: 1
> |  + Forced flag: 0
> |  + MinCache: 0
> |  + Max BlockAddition ID: 0
> |  + Codec decode all: 1
100a127,131
> |  + Enabled: 1
> |  + Forced flag: 0
> |  + MinCache: 0
> |  + Max BlockAddition ID: 0
> |  + Codec decode all: 1
113a145,149
> |  + Enabled: 1
> |  + Forced flag: 0
> |  + MinCache: 0
> |  + Max BlockAddition ID: 0
> |  + Codec decode all: 1
126a163,167
> |  + Enabled: 1
> |  + Forced flag: 0
> |  + MinCache: 0
> |  + Max BlockAddition ID: 0
> |  + Codec decode all: 1
136c177
< |  + Name: Oldboy
---
> |  + Name: Stereo
139a181,185
> |  + Enabled: 1
> |  + Forced flag: 0
> |  + MinCache: 0
> |  + Max BlockAddition ID: 0
> |  + Codec decode all: 1
152a199,203
> |  + Enabled: 1
> |  + Forced flag: 0
> |  + MinCache: 0
> |  + Max BlockAddition ID: 0
> |  + Codec decode all: 1
161c212,216
< | + EbmlVoid (size: 35)
---
> |  + Enabled: 1
> |  + Forced flag: 0
> |  + MinCache: 0
> |  + Max BlockAddition ID: 0
> |  + Codec decode all: 1
170c225,229
< | + EbmlVoid (size: 35)
---
> |  + Enabled: 1
> |  + Forced flag: 0
> |  + MinCache: 0
> |  + Max BlockAddition ID: 0
> |  + Codec decode all: 1
179c238,242
< | + EbmlVoid (size: 35)
---
> |  + Enabled: 1
> |  + Forced flag: 0
> |  + MinCache: 0
> |  + Max BlockAddition ID: 0
> |  + Codec decode all: 1
188c251,256
< | + EbmlVoid (size: 35)
---
> |  + Enabled: 1
> |  + Forced flag: 0
> |  + MinCache: 0
> |  + Max BlockAddition ID: 0
> |  + Codec decode all: 1
> |+ EbmlVoid (size: 903)
```

**Comment by mbunkus** [comment]
That's correct, but unfortunately kind of necessary due to the architecture of the libraries used under the hood. These additional header fields don't change the semantics as they're added with their default values (meaning a player must behave the same whether they're present in the file with that value of if they're not present at all).

I won't fix this. Sorry.

**Comment by enonu** [comment]
Thanks for the quick reply. What libraries underneath are causing the problem? Do you attempt a read of these fields from the library, and they're returning default values regardless of presence?

**Comment by mbunkus** [comment]
libMatroska.

libmatroska has two modes when writing a master element (a master element is something like the track headers):

1. Write all known elements, even if set to their default values,
2. Write only those elements that the program has told me to write but leave out all those that equal their default value.

Technically it's more complicated, but it boils down to these two choices.

Way back when I used to use the other method. Unfortunately this caused real problems with a lot of players (especially hardware devices which are known not to receive fixes anymore shortly after initial release) which didn't implement the Matroska specs properly. Therefore I changed the handling to the first method. It simply has fewer drawbacks in real world usage.
